include/forcing/GenericDataProvider.hpp

```diff
@@ -4,14 +4,26 @@
 #include "DataProvider.hpp"
 #include "DataProviderSelectors.hpp"
 
+#ifdef NETCDF_ACTIVE
+#include <memory>
+#include <netcdf>
+#endif
+
 namespace data_access
 {
     class GenericDataProvider : public DataProvider<double, CatchmentAggrDataSelector>
     {
         public:
 
+  #ifdef NETCDF_ACTIVE
+        /** Retrun the shared_ptr to the NcFile */
+        virtual std::shared_ptr<netCDF::NcFile> get_nc_file() {
```

Comment:
We should not have a NetCDF-specific function of any kind in GenericDataProvider.

Reply:
Does the guard isolates the get_nc_file() function while still keeps the GenericDataProvider's generality? The reason I had it there is because a GenericDataProvider type of pointer is used in Formulation_Constructor.hpp to point to a NetCDFPerFeatureDataProvider. It seemed this not the case looking at the code. But both gdb debugger and output seem to say so.

Reply:
No it doesn't keep the generality, as I think you're finding out when the build fails to find the .h file here... since it's in there at all, now there's a dependency on NetCDF introduced in this class, which has to be resolved in the build system. More generally, this is an abstract base class and shouldn't know anything about the details of its various subclass implementations.

Reply:
I don't have any good ideas at the moment how to by pass it. Will think about it.

Reply:
Ultimately, the issue here is injecting a netcdf specific function into the data provider interface.  So even if we have other provider types supported (e.g. CSV) the interface now suddenly includes a function that isn't relevant...also the macro guard makes the interface confusing in that general sense -- hence why this type of functionality is better suited to be encapsulated by the netcdf provider class.  I'll dig a little deeper into this PR and see if I can help with any ideas.

Reply:
Seems like @mattw-nws has done the latter part already in #485 .  It isn't quite as explicit, but that solution does two things:
Isolates netcdf specific semantics to the netcdf provider class.
Leverages default destruction semantics to ensure destructors are called appropriately on the netcdf pointers (which in turn trigger destruction/cleanup of the underlying hdf5 resource...).

Ultimately, the problem to solve here is ensure resource management in the right context and the right point in the object life cycle.  C++ gives you a lot of good and bad ways to do that, haha.